src/tls/server.c

```diff
@@ -0,0 +1,902 @@
+}
+
+static int
+gnutls_send_all (gnutls_session_t session, const char *data, size_t data_size)
+{
+  int r;
+
+  while (data_size > 0)
+  {
+    r = gnutls_record_send (session, data, data_size);
+
+    if (r == GNUTLS_E_AGAIN || r == GNUTLS_E_INTERRUPTED)
+      continue;
+    if (r < 0)
+      return r;
+    if (r == 0) /* Should Not Happen™, as data_size > 0 */
+      return GNUTLS_E_PUSH_ERROR;
+
+    data += r;
+    data_size -= r;
+  }
+  return 0;
+}
+
+/**
+ * handle_connection_data: Handle event on client or ws fd
+ *
+ * We want to avoid any interpretation of data to avoid vulnerabilities, so for
+ * the most part this just means shovelling data between the client and ws. The
+ * only exception is the very first byte of a new connection, to tell apart TLS
+ * from plain HTTP (handled by handle_connection_data_first).
+ */
+static void
+handle_connection_data (int fd)
+{
+  enum DataSource src;
+  Connection *con;
+  int ret;
+  char buffer[1024 * 1024];
+
+  con = get_fd_connection (fd, &src);
+  assert (con);
+  debug ("%s connection fd %i has data from %s; ws %s",
+         con->is_tls ? "TLS" : "unencrypted", con->client_fd,
+         src == WS ? "ws" : "client",
+         con->ws ? con->ws->socket.sun_path : "uninitialized");
+
+  /* first data on a new connection; determine if TLS, init TLS, and assign a ws */
+  if (!con->ws)
+    {
+      assert (src == CLIENT);
+      handle_connection_data_first (fd, con);
+      return;
+    }
+
+  /* TLS connections */
+  if (con->is_tls)
+    {
+      if (src == CLIENT)
+        {
+          TLS_RETRY (ret, gnutls_record_recv (con->session, buffer, sizeof (buffer)));
+          if (ret == 0)
+            {
+              debug ("peer has closed the TLS connection");
+              TLS_RETRY_BLOCK (ret, gnutls_bye (con->session, GNUTLS_SHUT_WR));
+              remove_connection (fd, NULL);
+            }
+          else if (ret < 0)
+            {
+              if (gnutls_error_is_fatal (ret))
+                {
+                  warnx ("%s", gnutls_strerror (ret));
+                  remove_connection (fd, NULL);
+                }
+            }
+          else
+            {
+              int r = send_all (con->ws_fd, buffer, ret, 0);
+              if (r < 0)
+                {
+                  warn ("failed to send data");
+                  remove_connection (con->client_fd, NULL);
+                }
+            }
+        }
+      else /* src == WS */
+        {
+          RETRY (ret, recv (con->ws_fd, buffer, sizeof (buffer), MSG_DONTWAIT));
+          if (ret == 0)
+            {
+              debug ("peer has closed the connection");
+              remove_connection (con->client_fd, NULL);
+            }
+          else if (ret < 0)
+            {
+              warn ("failed to receive data");
+              remove_connection (con->client_fd, NULL);
+            }
+          else
+            {
+              int r = gnutls_send_all (con->session, buffer, ret);
```

Comment:
Same here re blocking writes.

Reply:
In addition, writing TLS data might actually need to read from the peer during a renegotiation (which might happen at any time, I think), so blocking here until everything is written is considerably simpler.

Reply:
To make sure I understand your second comment: TLS renegotiation is opaque to the GnuTLS API, so if that happens in the background, `gnutls_record_send()` needs to deal with that internally. With the approach outlined above, I'd put the socket in non-blocking mode, so that in the event of a full fd buffer, it would fail with EAGAIN, and then just get flushed out at the next opportunity. The main difficulty here is that while the browser tries to send something, but we ignore EPOLLIN as we still have data to write, cockpit-tls might never reattempt the _record_send(), and we run into a deadlock?

So you are saying that even with the approach above, in this one  case we should still use `gnutls_send_all()` (which blocks on the renegotiation), and effectively ignore epoll during that time? That seems like a robust approach, even though a broken or rogue browser could then DoS cockpit-tls.

Reply:
> The main difficulty here is that while the browser tries to send something, but we ignore EPOLLIN as we still have data to write, cockpit-tls might never reattempt the _record_send(), and we run into a deadlock?

Yes.

Reply:
So for the time being, we could make the Unix socket to cockpit-ws and http socket to browser non-blocking, but keep the https socket to the browser as blocking for this. Reads will still be non-blocking.